drivers/src/stm32f446xx_usart_driver_course.c

```diff
@@ -110,3 +153,63 @@ void USART_WriteData(USART_Handler_TypedDef* pUSART_Handler, uint8_t* pUSART_TX_
 	{
 	}
 }
+
+void USART_SetBaudRate(USART_Handler_TypedDef* pUSART_Handler)
+{
+	uint16_t tempreg, mantissa, DIV_fraction, DIV_mantissa;
+	float fraction;
+
+	uint32_t usart_clock, usart_div;
+
+	//UART1 and USART6 are on APB2 so they use PCLK2
+	if((pUSART_Handler->pUSART == UART1) || (pUSART_Handler->pUSART == UART6))
+	{
+		usart_clock = RCC_GetPCLK2();
+	}
+	else
+	{
+		usart_clock = RCC_GetPCLK1();
+	}
+
+	usart_div = (usart_clock*100)/((8*2-pUSART_Handler->USART_Pin_Config.USART_OverSampling)*pUSART_Handler->USART_Pin_Config.USART_BaudRate);
+
+	mantissa = usart_div/100;
+	fraction = usart_div - (mantissa*100);
+	fraction = fraction/100;
+
+	if(pUSART_Handler->USART_Pin_Config.USART_OverSampling == 0)	//oversampling = 16, fraction coded on 4 bits
+	{
+		DIV_fraction = (fraction*16) + 0.5; //round to the nearest number
+		if(DIV_fraction > 0xF)	//test to see if the fraction bits overflow
+		{
+			DIV_mantissa = mantissa + 1;
+			DIV_fraction = 0;
+		}
+		else
+		{
+			DIV_mantissa = mantissa;
+		}
+
+	}
+	else if(pUSART_Handler->USART_Pin_Config.USART_OverSampling == 1)	//oversampling = 8, fraction coded on 3 bits, fraction bit 3 must be left cleared
+	{
+		DIV_fraction = (fraction*8) + 0.5;	//round to the nearest number
+
+		if(DIV_fraction > 0x7)	//test to see if the fraction bits overflow
+		{
+			DIV_mantissa = mantissa + 1;
+			DIV_fraction = 0;
+		}
+		else
+		{
+			DIV_mantissa = mantissa;
+		}
+	}
+	tempreg &= ~(0x0000FFFF);
+	pUSART_Handler->pUSART->BRR &= tempreg; //bits 0-3 are fraction, bits 4-15 are mantissa
+
+	tempreg |= (mantissa << 4);
+	tempreg |= (fraction);
```

Comment:
Float into a register?

Reply:
My bad, I forgot to change it to "DIV_fraction'